pkg/action/mount.go

```diff
@@ -0,0 +1,340 @@
+		cfg.Logger.Errorf("Error mounting image %s: %s", spec.Image.File, err.Error())
+		return err
+	}
+
+	cfg.Logger.Debugf("Writing fstab")
+	if err := WriteFstab(cfg, spec); err != nil {
+		cfg.Logger.Errorf("Error writing new fstab: %s", err.Error())
+		return err
+	}
+
+	cfg.Logger.Info("Mount command finished successfully")
+	return nil
+}
+
+func applyLayoutConfig(cfg *v1.RunConfig, spec *v1.MountSpec) error {
+	// Read the OVERLAY, RW_PATHS, PERSISTENT_STATE_PATHS and PERSISTENT_STATE_BIND and overwrite the MountSpec
+	files := []string{"/run/cos/cos-layout.env", "/run/elemental/layout.env"}
+
+	for _, file := range files {
+		cfg.Logger.Debugf("Parsing env vars from file '%s'", file)
+		env, err := godotenv.Read(file)
+		if err != nil {
+			cfg.Logger.Warnf("Failed reading file %s: %s", file, err.Error())
+			continue
+		}
+
+		if overlay, exists := env["OVERLAY"]; exists {
+			cfg.Logger.Debug("Found OVERLAY env var")
+
+			split := strings.SplitN(overlay, ":", 2)
+
+			if split[0] == "tmpfs" && len(split) == 2 {
+				spec.Overlay.Size = split[1]
+			}
+		}
+
+		if rwPaths, exists := env["RW_PATHS"]; exists {
+			cfg.Logger.Debug("Found RW_PATHS env var")
+			spec.Overlay.Paths = strings.Fields(rwPaths)
+		}
+
+		if paths, exists := env["PERSISTENT_STATE_PATHS"]; exists {
+			cfg.Logger.Debug("Found PERSISTENT_STATE_PATHS env var")
+			spec.Persistent.Paths = strings.Fields(paths)
+		}
+
+		if _, exists := env["PERSISTENT_STATE_BIND"]; exists {
+			cfg.Logger.Debug("Found PERSISTENT_STATE_BIND env var")
+			spec.Persistent.Mode = constants.BindMode
+		}
+	}
+
+	return nil
+}
+
+func RunFsck(cfg *v1.RunConfig, spec *v1.MountSpec, e *elemental.Elemental) error {
+	if !spec.RunFsck {
+		cfg.Logger.Debug("Skipping fsck")
+		return nil
+	}
+
+	allParts, err := utils.GetAllPartitions()
+	if err != nil {
+		cfg.Logger.Errorf("Error getting all partitions: %s", err.Error())
+		return err
+	}
+
+	return e.FsckPartitions(allParts)
+}
+
+func MountOverlay(cfg *v1.RunConfig, sysroot string, overlay v1.OverlayMounts) error {
+	if err := utils.MkdirAll(cfg.Config.Fs, constants.OverlayDir, constants.DirPerm); err != nil {
+		cfg.Logger.Errorf("Error creating directory %s: %s", constants.OverlayDir, err.Error())
+		return err
+	}
+
+	var (
+		overlaySource string
+		overlayFS     string
+		overlayOpts   []string
+	)
+
+	switch overlay.Type {
+	case constants.Tmpfs:
+		overlaySource = constants.Tmpfs
+		overlayFS = constants.Tmpfs
+		overlayOpts = []string{"defaults", fmt.Sprintf("size=%s", overlay.Size)}
+	case constants.Block:
+		overlaySource = overlay.Device
+		overlayFS = constants.Autofs
+		overlayOpts = []string{"defaults"}
+	default:
+		return fmt.Errorf("unknown overlay type '%s'", overlay.Type)
+	}
+
+	if err := cfg.Mounter.Mount(overlaySource, constants.OverlayDir, overlayFS, overlayOpts); err != nil {
+		cfg.Logger.Errorf("Error mounting overlay: %s", err.Error())
+		return err
+	}
+
+	for _, path := range overlay.Paths {
```

Comment:
a small issue we encountered was bind paths not being ordered, so something like this mistake in the layout file:

```
        /usr/share/pki/trust/anchors
	/usr/share/pki/trust
```

would shadow the first bind with the second! Just a heads up as we encountered this :) Patch looks awesome :+1:

Reply:
@Itxaka thanks for the heads up! I will add a commit to sort these in correct order! :+1: